cmd/github-ldap-user-group-creator/main.go

```diff
@@ -142,12 +148,46 @@ func main() {
 		logrus.WithError(err).Fatal("Failed to load the mapping")
 	}
 
-	if err := ensureGroups(ctx, clients, mapping, opts.dryRun); err != nil {
+	data, err := ioutil.ReadFile(opts.groupsFile)
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to read the group file")
+	}
+	roverGroups := map[string][]string{}
+	if err := yaml.Unmarshal(data, &roverGroups); err != nil {
+		logrus.WithError(err).Fatal("Failed to unmarshal groups")
+	}
+
+	if ciAdmins, ok := roverGroups[api.CIAdminsGroupName]; !ok {
+		logrus.WithField("groupName", api.CIAdminsGroupName).Fatal("Failed to find ci-admins group")
+	} else if l := len(ciAdmins); l < 3 {
+		logrus.WithField("groupName", api.CIAdminsGroupName).WithField("len", l).Fatal("Require at least 3 members of ci-admins group")
+	}
+
+	if err := ensureGroups(ctx, clients, mapping, roverGroups, opts.dryRun); err != nil {
 		logrus.WithError(err).Fatal("could not ensure groups")
 	}
 }
 
-func ensureGroups(ctx context.Context, clients map[string]ctrlruntimeclient.Client, mapping map[string]string, dryRun bool) error {
+func ensureGroups(ctx context.Context, clients map[string]ctrlruntimeclient.Client, mapping map[string]string, roverGroups map[string][]string, dryRun bool) error {
+
+	groupsNamesToCreate := sets.NewString()
+	var groupsToCreate []userv1.Group
+	for githubLogin, kerberosId := range mapping {
+		groupName := fmt.Sprintf("%s%s", githubLogin, api.GroupSuffix)
+		groupsNamesToCreate.Insert(groupName)
+		groupsToCreate = append(groupsToCreate, userv1.Group{
+			ObjectMeta: metav1.ObjectMeta{Name: groupName, Labels: map[string]string{api.DPTPRequesterLabel: toolName}},
+			Users:      userv1.OptionalNames{githubLogin, kerberosId},
+		})
+	}
+	for k, v := range roverGroups {
+		groupsNamesToCreate.Insert(k)
+		groupsToCreate = append(groupsToCreate, userv1.Group{
+			ObjectMeta: metav1.ObjectMeta{Name: k, Labels: map[string]string{api.DPTPRequesterLabel: toolName}},
+			Users:      v,
+		})
+	}
```

Comment:
I think that this method ends up doing too much: figuring out which groups should exist from two separate sources, and then creating them.

How about splitting the three actions into three separate methods, each with much simpler interface, called straight from the `main()` level? Somehting like:

```go
func makeGitHubMappingGroups(mapping) map[string]*userv1.Group {...}
func makeRoverGroups(roverGroups) map[string]*userv1.Group {...}

func ensureGroups(ctx, clients, dryRun, groups map[string]*userv1.Group) error {
  groupNames := sets.NewString()
  for name := groups {
    groupNames.Insert(name)
  }
  ...
}

func main() {
...
  createGroups := map[string]*userv1.Group
  // iterate over sources of groups to create
  for groups := range []map[string]*userv1.Group { makeGitHubMappingGroups(mappings), makeRoverGroups(roverGroups) } {
    for name, group := range groups {
      if _, exists := createGroups[name]; exists { 
        // error out, duplicate group in rover and mapping
      }
      createGroups[name] = group 
    }
  }
}
```

Reply:
Done.